---
title: "scRNA_Seq_Seurat_Tutorial"
output: html_document
date: "2024-07-21"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## This document walk you through single cell RNA-Seq analysis using various R Packages

This analysis utilizes sample files from unpublished study GSE256291, processed with 10x Genomics protocol. Data is used for demonstration purposes only.



```{r loading_packages}
###Loading Libraries
library(rmarkdown)
library(remotes)
library(Seurat)
library(dplyr)
library(presto)
library(devtools)
library(patchwork)
library(ggplot2)

```

## Loading datasets

The initial step involves loading datasets to generate a Seurat object, which consolidates all samples for downstream analysis.
For 10x technology data, three files per sample (barcodes, features, and matrix) must reside in a single directory for the Read10X() function. Otherwise, individual loading is required.

```{r , echo=FALSE}
patinet1_counts <- Read10X(data.dir = "~/Github_Repos/inputs/data/patient1/")
patinet2_counts <- Read10X(data.dir = "~/Github_Repos/inputs/data/patient2/")
healthy1_counts <- Read10X(data.dir = "~/Github_Repos/inputs/data/healthy1/")
healthy2_counts <- Read10X(data.dir = "~/Github_Repos/inputs/data/healthy2/")
```

Individual Seurat objects are created and subsequently combined using the merge() function.


```{r }
seurat1 <- CreateSeuratObject(patinet1_counts, project="cd14_patient1")
seurat2 <- CreateSeuratObject(patinet2_counts, project="cd14_patient2")
seurat3 <- CreateSeuratObject(healthy1_counts, project="cd14_healthy1")
seurat4 <- CreateSeuratObject(healthy2_counts, project="cd14_healthy2")
cd14_combined <- merge(seurat1, y=c(seurat2,seurat3,seurat4), add.cell.ids = c("patient1", "patient2","healthy1","healthy2"), project = "cd14")
```

After object creation, we analyze cellular RNA distribution by calculating mitochondrial RNA percentages and visualizing them collectively. The data can be displayed with or without data points.

```{r,fig.width=9,fig.height=5}
cd14_combined[["percent.mt"]] <- PercentageFeatureSet(cd14_combined, pattern = "^MT[-\\.]")
VlnPlot(cd14_combined, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
```



```{r,fig.width=9,fig.height=5}
VlnPlot(cd14_combined, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3, pt.size=0)
```

We examine the correlation between transcript and gene counts, as well as between mitochondrial features and RNA quantity. The expected patterns are a positive correlation between feature and RNA counts, and a negative correlation between mitochondrial features and RNA quantity.


```{r,fig.width=11}
plot1 <- FeatureScatter(cd14_combined, feature1 = "nCount_RNA", feature2 = "percent.mt",pt.size=1)
plot2 <- FeatureScatter(cd14_combined, feature1 = "nCount_RNA", feature2 = "nFeature_RNA",pt.size=1)
plot1 + plot2
```

Following QC plot analysis, we filter cells and normalize the data for comparative analysis. The 'LogNormalize' method normalizes each cell's feature expression by total expression, scaled by a default factor of 10,000.


```{r}
cd14_combined <- subset(cd14_combined, subset = nFeature_RNA > 500 & nFeature_RNA < 7000 & percent.mt < 5)
cd14_combined <- NormalizeData(cd14_combined)
```

We identify the most variable genes using a threshold of 2,000 features. While typical ranges are 2,000-5,000, the optimal number depends on the dataset and research objectives.


```{r}
cd14_combined <- FindVariableFeatures(cd14_combined, nfeatures = 2000)
```


Now, we can visualize the result

```{r,fig.width=10,fig.height=4}
top_features <- head(VariableFeatures(cd14_combined), 20)
plot1 <- VariableFeaturePlot(cd14_combined)
plot2 <- LabelPoints(plot = plot1, points = top_features, repel = TRUE)
plot1 + plot2
```

Gene expression data requires scaling to prevent highly expressed genes from dominating the analysis. This normalization step ensures equal weighting of genes regardless of their baseline expression levels.


```{r}
cd14_combined <- ScaleData(cd14_combined)
cd14_combined <- ScaleData(cd14_combined, vars.to.regress = c("nFeature_RNA", "percent.mt"))
```


We perform dimensionality reduction using PCA (default 50 principal components), followed by UMAP and t-SNE visualization in 2D space.



```{r}

cd14_combined <- RunPCA(cd14_combined, npcs = 50)

ElbowPlot(cd14_combined, ndims = ncol(Embeddings(cd14_combined, "pca")))
          
```


Here we can visualize which gene is contributing to which PC (top 4)

```{r,fig.width=12,fig.height=12}

PCHeatmap(cd14_combined, dims = 1:5, cells = 500, balanced = TRUE, ncol = 5)

```

Another method for visualization of the PCA analysis.

```{r, fig.height=10}
VizDimLoadings(cd14_combined, dims = 1:2, reduction = "pca")
```

Another method

```{r}
DimPlot(cd14_combined, reduction = "pca") + NoLegend()
cd14_combined <- RunTSNE(cd14_combined, dims = 1:20)
cd14_combined <- RunUMAP(cd14_combined, dims = 1:20)
```
```{r, fig.width=12,fig.height=6}
plot1 <- TSNEPlot(cd14_combined)
plot2 <- UMAPPlot(cd14_combined)
plot1 + plot2
```


Based on the visualization results, t-SNE reveals more distinct sample clustering than other methods. We now select 5 genes from our previously identified highly variable genes.


```{r, fig.width=14,fig.height=5}
top_features
plot1 <- FeaturePlot(cd14_combined, c("FCGR3A","CCDC26","ETS1","ITK","SEL1L3"),
                     ncol=5, reduction = "tsne")

plot1

```

Clustering begins by generating a k-nearest neighbor network based on PC values. Cells are connected to their nearest neighbors, with connections weighted by shared neighbor proportions. Weak connections are removed to create the final Shared Nearest Neighbor (SNN) network.


```{r}
cd14_combined <- FindNeighbors(cd14_combined)
```


The Louvain algorithm is applied to identify cell communities within the network, detecting groups with dense internal connections but sparse connections between different groups.


```{r}
cd14_combined <- FindClusters(cd14_combined, resolution = 1)

```

```{r, fig.width=12}

plot1 <- DimPlot(cd14_combined, reduction = "tsne", label = TRUE)
plot2 <- DimPlot(cd14_combined, reduction = "umap", label = TRUE)
plot1 + plot2
```

The Harmony package offers methods for batch effect correction, which may be necessary when batch effects are present in the data.


```{r}
library(harmony)

cd14_combined <- RunHarmony(cd14_combined, group.by.vars = "orig.ident", dims.use = 1:20, max.iter.harmony = 50)
cd14_combined <- RunUMAP(cd14_combined, reduction = "harmony", dims = 1:20)
cd14_combined <- FindNeighbors(cd14_combined, reduction = "harmony", dims = 1:20) %>% FindClusters(resolution = 0.6)

DimPlot(cd14_combined, reduction = "harmony", label = F)


```


Here, We identify and filter top marker genes for each identified cluster.


```{r}
cd14_combined <- JoinLayers(cd14_combined)
cd14_markers <- FindAllMarkers(cd14_combined, only.pos = TRUE, min.pct = 0.25, logfc.threshold = log(1.2))
top_marker <- cd14_markers %>% group_by(cluster) %>% top_n(n = 1, wt = avg_log2FC)
top_marker
```

Now we can visualize the result

```{r}
DoHeatmap(cd14_combined, features = top_marker$gene) + NoLegend()
```

Cell annotation, a critical step in single-cell analysis, involves manual annotation of clusters using identified markers and validation through the panglaodb.se database, a comprehensive single-cell RNA-Seq reference.

```{r}
new_cluster_names <- c("Macrophages",
                        "Fibroblasts",
                        "Macrophages",
                        "T Cells",
                        "Keratinocytes",
                        "Monocytes",
                        "Fibroblasts",
                        "Dendritic cells",
                        "Unknown",
                        "Epithelial Cells",
                        "T Cells",
                        "Fibroblasts",
                        "T Cells",
                        "B Cells")

names(new_cluster_names) <- levels(cd14_combined)
cd14_combined <- RenameIdents(cd14_combined, new_cluster_names)
DimPlot(cd14_combined, reduction = "tsne", label = TRUE)
```


Non-negative Matrix Factorization (NMF) clustering helps identify patterns in complex data. We apply this dimensionality reduction method to both UMAP and t-SNE visualizations.


```{r}
library(GeneNMF)
cd14_combined <- runNMF(cd14_combined, k = 13, assay="RNA")

cd14_combined <- RunTSNE(cd14_combined, reduction = "NMF", dims=1:13, reduction.name = "NMF_Tsne", reduction.key = "nmfTsne_")
cd14_combined <- RunUMAP(cd14_combined, reduction = "NMF", dims=1:13, reduction.name = "NMF_UMAP", reduction.key = "nmfUMAP_")

```

The results can now be visualized using both UMAP and t-SNE representations.

```{r}
DimPlot(cd14_combined, reduction = "NMF_Tsne", label=F) + theme(aspect.ratio = 1,
                                                            axis.text = element_blank(),
    axis.title = element_blank(),
    axis.ticks = element_blank()) + ggtitle("NMF TSNE")
```

```{r}
DimPlot(cd14_combined, reduction = "NMF_UMAP", label=F) + theme(aspect.ratio = 1,
                                                            axis.text = element_blank(),
    axis.title = element_blank(),
    axis.ticks = element_blank()) + ggtitle("NMF UMAP")
```

NMF analysis reveals 10 distinct clusters, in contrast to Seurat's 13 clusters. We can now examine the most variable genes within each NMF-derived cluster.


```{r}
cd14_list <- SplitObject(cd14_combined, split.by = "orig.ident")
cd14_multi <- multiNMF(cd14_list, assay="RNA", slot="data", k=4:9, nfeatures = 1000)
cd14_meta <- getMetaPrograms(cd14_multi,
                                        nMP=10,
                                        weight.explained = 0.7,
                                        max.genes=100)
NMF_Heatmap <- plotMetaPrograms(cd14_meta)
```

We now examine each component (cluster) in detail.


```{r}
lapply(cd14_meta$metaprograms.genes, head)
```

```{r,fig.width=11}
library(UCell)
mp.genes <- cd14_meta$metaprograms.genes
cd14_combined <- AddModuleScore_UCell(cd14_combined, features = mp.genes, assay="RNA", ncores=4, name = "")
VlnPlot(cd14_combined, features=names(mp.genes),
        pt.size = 0, ncol=6)
```



```{r}
sessionInfo()
```
